### PR TITLE
Feat: Choose Directory and add custom path

### PR DESCRIPTION
**Pull request by 0tuedon**
https://github.com/bitcointranscripts/transcription-review-front-end/assets/90271995/163dab30-e7e9-4a68-aa2e-e00f7660394d

1. You can be able to select a directory from the root
2. You can create a custom path

PENDING
1. No nested directories yet

### CODE REVIEW

**Comment by 0tuedon** [comment]
#105

**Comment by 0tuedon** [comment]
https://github.com/bitcointranscripts/transcription-review-front-end/assets/90271995/7cddd9b0-1bd1-47e7-b8c0-fd69dcf2ae22

This was quite challenging mapping out and showing options based on where they are with the directories, @Emmanuel-Develops , i also made sure it does make a request when the nestDir has been saved.
I think this is ready for review

File: src/components/sideBarContentEdit/SelectField.tsx [hunk 1]
```diff
@@ -15,6 +16,17 @@ type Props = {
   userCanAddToList?: boolean;
 };
 
+type PropsDirectory = {
+  name: string;
+  editedData: string;
+  index: number;
+  path: string;
+  // eslint-disable-next-line no-unused-vars
```
Comment (Extheoisah) [inline src/components/sideBarContentEdit/SelectField.tsx:24]: Why do we have this eslint? Is this prop not being used?

File: src/components/sideBarContentEdit/SelectField.tsx [hunk 2]
```diff
@@ -293,3 +305,85 @@ export const SingleSelectField = ({
...
+    return []; // Return null if the desired depth is reached, but the path is not found.
+  }
+
+  if (index === 1) {
+    return objArray;
+  }
+
+  for (const obj of objArray) {
+    if (obj.slug === path[depth - 2]) {
+      if (path.length === 2 && depth === 2) {
```
Comment (Extheoisah) [inline src/components/sideBarContentEdit/SelectField.tsx:327]: Can you put a comment on why we're doing this before the recursion? Its entirely clear

File: src/components/sideBarContentEdit/SidebarContentEdit.tsx [hunk 3]
```diff
@@ -38,7 +70,45 @@ const SidebarContentEdit = ({
...
+  const [directoryList, setDirectoryList] = useState<IDir[] | []>([]);
+  useEffect(() => {
+    // we want the rootpath to load first before
+    if (directoryPaths && initialCount < 2) {
+      setPath(`${data.content.loc}/`);
+      setInitialCount(2);
+    }
+  }, [data.content.loc, directoryPaths, initialCount]);
+  useEffect(() => {
+    const convertDirStructure = (
```
Comment (Extheoisah) [inline src/components/sideBarContentEdit/SidebarContentEdit.tsx:86]: Does this create a new mutated instance ofthe directiory or it duplicates it?

File: src/pages/api/github/dir.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import { Octokit } from "@octokit/core";
+import type { NextApiRequest, NextApiResponse } from "next";
+import { getSession } from "next-auth/react";
+
+async function getAllRepoFolder(
+  octokit: InstanceType<typeof Octokit>,
+  path?: string
+) {
+  const auth = await octokit.request("GET /user");
+  // Fork the repository
```
Comment (Extheoisah) [inline src/pages/api/github/dir.ts:11]: Remove this comment. This doesn't fork the repo

File: src/components/sideBarContentEdit/SelectField.tsx (hunk 1 above)
Comment (0tuedon) [reply src/components/sideBarContentEdit/SelectField.tsx:24]: > Why do we have this eslint? Is this prop not being used?

It is being used but on line 25 the parameter is not being return and esLint throws a warning for that

File: src/components/sideBarContentEdit/SelectField.tsx (hunk 2 above)
Comment (0tuedon) [reply src/components/sideBarContentEdit/SelectField.tsx:327]: > Can you put a comment on why we're doing this before the recursion? Its entirely clear
for line 327? 
reason for doing , we want to check the current length of the array for example advancing-bitcoin/2019 would be of length 2,  and if path if found we return the nestDir or an empty array if it's not two we slice it again

File: src/components/sideBarContentEdit/SidebarContentEdit.tsx (hunk 3 above)
Comment (0tuedon) [reply src/components/sideBarContentEdit/SidebarContentEdit.tsx:86]: > Does this create a new mutated instance ofthe directiory or it duplicates it?

it creates a mutated instance.

File: src/pages/api/github/dir.ts (hunk 4 above)
Comment (0tuedon) [reply src/pages/api/github/dir.ts:11]: Sure

**Comment by 0tuedon** [comment]
> Left a few comments. I need to test this pr and give feedback

Noted

File: src/components/sideBarContentEdit/SelectField.tsx [hunk 5]
```diff
@@ -293,3 +305,85 @@ export const SingleSelectField = ({
...
+    index + 1
+  );
+  const [value, setValue] = useState<string>(editedData);
+  const handleAutoCompleteSelect = (data: AutoCompleteData) => {
+    handleAddItem(data.value);
+    setValue(data.value);
+  };
+
+  // remove previously selected option from list
+  const newAutoCompleteList =
```
Comment (Emmanuel-Develops) [inline src/components/sideBarContentEdit/SelectField.tsx:372]: This is for selecting multiple items. Selecting location is single use so there's no need to remove previous items from the list

**Comment by adamjonas** [comment]
This isn't ready. Tried adding a new dir a couple of different ways.

https://github.com/bitcointranscripts/transcription-review-front-end/assets/755825/496c4870-1254-4d0d-91e5-c2d2fe5a8d3a

https://github.com/bitcointranscripts/transcription-review-front-end/assets/755825/630363b0-93f4-4497-a5b6-48968381b57b

**Comment by 0tuedon** [comment]
https://github.com/bitcointranscripts/transcription-review-front-end/assets/90271995/b9707b19-92f0-4644-a1b2-3cf75271feb4

A new look on the select directories flow @Extheoisah  @Emmanuel-Develops  @BalogunofAfrica . Please review

**Comment by adamjonas** [comment]
> You can add a new directory, demo is below, i think it might be the choice of words or the way of adding it?

I see. Yeah, that's not very intuitive. I liked the add new dir prompt. We'll also want to add an "are you sure?" modal to discourage people from doing this often (but not needed in a v1).

**Comment by Emmanuel-Develops** [comment]
@adamjonas we are taking a different approach with the logic for selection
I have a figma design [here](https://www.figma.com/file/82tn8JtimD4nYmAkMo85hM/Nested-select-UI?type=design&node-id=0%3A1&mode=design&t=g96mzuneG4ASIqum-1)
Showing a simple structure where users can easily navigate directories. As well as create arbitrary ones.
@0tuedon has complete freedom over the bits and pieces of its implementation

My mistake for not sharing the document earlier so the team could critic the doc for improvements

**Comment by 0tuedon** [comment]
> > You can add a new directory, demo is below, i think it might be the choice of words or the way of adding it?
> 
> I see. Yeah, that's not very intuitive. I liked the add new dir prompt. We'll also want to add an "are you sure?" modal to discourage people from doing this often (but not needed in a v1).
> 
> Why did you stray from the previous design? I thought it functioned fine, it just had saving issues.

for the confirmation modal, i would be adding it. 
Also  we took a different approach from the previous designs. reason being it create and new instance of the dropdown for each selection.

**Comment by 0tuedon** [comment]
> > > @0tuedon not all the directories loading?
> > > e.g. chaincode-labs has no sub-directories to select.
> > > Removing the ability to add a new directory is a loss. I'm open to it, but I preferred the old way of adding a new dir.
> > 
> > 
> > You can add a new directory, demo is below, i think it might be the choice of words or the way of adding it?
> > Screen.Recording.2023-07-25.at.16.43.41.mov
> > so when a directory is selected it saves and updates the transcript before getting the nested directories
> > Screenshot 2023-07-25 at 16 46 03
> 
> Looks good! a couple of comments on this:
> 
> * the back button is only visible if you're not in the root directory
> * the text next to the back button can be more pronounced (try bold but grey) as that is the current path the user is in  delimited by "/" as they go further or the current dir the user is at (which is the implementation you have here) try either and see what makes the best UX
> * switch text in the last button from "select" to "add" if path doesn't exist (path from github/ tree you built in state). Adding this arbitrary path should have a confirmation popup.
> * also the back button duo-tone could be a bit distracting, try a single background or just the arrow
> * align the options ("chaincode-podcast", "chaincode-residency") vertically to the folder ("chaincode-labs"). This references the second point which helps separate current directory from subdirectories under it

Made an update to add confirmation modal and the listed minor changes
https://github.com/bitcointranscripts/transcription-review-front-end/assets/90271995/9156f27d-ff69-44f5-bf69-3966d924e31c

**Comment by 0tuedon** [comment]
@Emmanuel-Develops Prior to our discussion, I have been able to implement the changes request, please review

https://github.com/bitcointranscripts/transcription-review-front-end/assets/90271995/b0a6b9d8-42a0-4239-a811-a6360b0e0664

This video is for and found path


https://github.com/bitcointranscripts/transcription-review-front-end/assets/90271995/f0be07eb-0e4d-4b32-b4aa-e1d82bddca03

This is for the custom path

**Comment by adamjonas** [comment]
@0tuedon the commit history isn't right here. Can you rebase your changes on the dev branch and repush please?

**Comment by 0tuedon** [comment]
> @0tuedon the commit history isn't right here. Can you rebase your changes on the dev branch and repush please?

Alright

**Comment by 0tuedon** [comment]
> @0tuedon the commit history isn't right here. Can you rebase your changes on the dev branch and repush please?

i  think i have updated the branch and rebased this already. although i just made an update now.

**Comment by 0tuedon** [comment]
> @0tuedon the commit history isn't right here. Can you rebase your changes on the dev branch and repush please?

PR #145   fixes the commit history on this

**Pull request closed without merging**